### ISSUE TITLE
We need a way to set environment variables on a per-miner basis

### ISSUE THREAD

**Issue by grantemsley** [MultiPoolMiner/MultiPoolMiner#1834]
GPU_FORCE_64BIT_PTR is causing problems.  This mostly seems to affect AMD cards.

Some miners *need* it set to 1.  Other *need* it set to 0.

I think we need to add another property to the miners for environment variables that need to be set, then when the job is created in the StartMining() function set those variables inside the job where they should only affect that one process.

**Comment by UselessGuru** [comment]
Also see https://github.com/MultiPoolMiner/MultiPoolMiner/issues/227

**Comment by UselessGuru** [comment]
> I think we need to add another property to the miners for environment variables that need to be set

In 3.30 I've added a property `Environment = @("GPU_FORCE_64BIT_PTR=0")` to the miners that need it.
The code used with `"ShowMinerWindow": true,` () makes usage of this:
```
                if ((Test-Path ".\CreateProcess.cs" -PathType Leaf) -and -not $this.Environment) {
                    $this.Process = Start-SubProcessWithoutStealingFocus -FilePath $this.Path -ArgumentList $this.GetCommandLineParameters() -WorkingDirectory (Split-Path $this.Path) -Priority ($this.DeviceName | ForEach-Object {if ($_ -like "CPU#*") {-2} else {-1}} | Measure-Object -Maximum | Select-Object -ExpandProperty Maximum)
                }

```
But the other launcher methods currently cannot make use of it.

Aaron: Any ideas?

**Comment by aaronsace** [comment]
Sounds like a bug with the miners. Can you not just patch the API ps1 files or do the miners all share the same API?

**Comment by UselessGuru** [comment]
@aaronsace 

> Can you not just patch the API ps1 files

All miner use this code (Include.psm1 / StartMining)
```
            if ($this.ShowMinerWindow -and -not ($this.API -eq "Wrapper") -or $this.Environment) {
                if ((Test-Path ".\CreateProcess.cs" -PathType Leaf) -and -not $this.Environment) {
                    $this.Process = Start-SubProcessWithoutStealingFocus -FilePath $this.Path -ArgumentList $this.GetCommandLineParameters() -WorkingDirectory (Split-Path $this.Path) -Priority ($this.DeviceName | ForEach-Object {if ($_ -like "CPU#*") {-2} else {-1}} | Measure-Object -Maximum | Select-Object -ExpandProperty Maximum)
                }
                else {
                    $this.Process = Start-Job ([ScriptBlock]::Create("Start-Process $(@{desktop = "powershell"; core = "pwsh"}.$Global:PSEdition) `"-command $EnvCmd```$Process = (Start-Process '$($this.Path)' '$($this.GetCommandLineParameters())' -WorkingDirectory '$(Split-Path $this.Path)' -WindowStyle Minimized -PassThru).Id; Wait-Process -Id `$PID; Stop-Process -Id ```$Process`" -WindowStyle Hidden -Wait"))
                }
            }
            else {
                $this.LogFile = $Global:ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath(".\Logs\$($this.Name)-$($this.Port)_$(Get-Date -Format "yyyy-MM-dd_HH-mm-ss").txt")
                $this.Process = Start-SubProcess -FilePath $this.Path -ArgumentList $this.GetCommandLineParameters() -LogPath $this.LogFile -WorkingDirectory (Split-Path $this.Path) -Priority ($this.DeviceName | ForEach-Object {if ($_ -like "CPU#*") {-2} else {-1}} | Measure-Object -Maximum | Select-Object -ExpandProperty Maximum)
            }

```
There are 3 ways to launch a miner process:

- $this.Process = Start-SubProcessWithoutStealingFocus
   (+) Does show miner windows, but does not steal focus
   (-) Does not work with Wrapper
   (-) does not allow env to be set;

- $this.Process = Start-Job
   (+) Does show miner windows
   (-) Does steal focus
   (-) Does not work with Wrapper
   (+) Does allow env to be set;

- $this.Process = Start-SubProcess
   (+) Works with wrapper
   (-) Does NOT show miner windows; does not steal focus
   (-) Does allow env to be set;

Visible miner windows are important to many users, and they should not lose focus.

**Comment by grantemsley** [comment]
@UselessGuru it should be possible to set enviroment variables with Start-SubProcessWithoutStealingFocus as well.  I've never tried it though.  The variable that controls the environment is $lpEnvironment.  Right now we have `$lpEnvironment = [IntPtr]::Zero` which tells it to just use the environment of the calling process (it's equivalent to null in this function call).

You'd need to build a pointer to a variable that has a null terminated list of environment variables.  eg. `GPU_FORCE_64BIT_PTR=1\0OTHERVARIABLE=STUFF\0`

I don't know how to do that, but it should be possible.

https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa#parameters look under lpEnvironment.

**Comment by UselessGuru** [comment]
@grantemsley 

Thanks for your input!

> I don't know how to do that, but it should be possible.

That is exactly my issue too. I'll see what I can do.

**Comment by UselessGuru** [comment]
> You'd need to build a pointer to a variable that has a null terminated list of environment variables. eg. > GPU_FORCE_64BIT_PTR=1\0OTHERVARIABLE=STUFF\0
>I don't know how to do that, but it should be possible.

No matter how much I try with 'lpEnvironment' I fail....

BUT I found an easier way to set the environment for Start-SubProcessWithoutStealingFocus.
It is sufficient to set the env variables in the function itself BEFORE starting the job. The job will then inherit the environment. Problem solved ;-)

This change will be included in the coming 3.3.0 Beta 20.
